src/main.rs

```diff
@@ -0,0 +1,292 @@
+//! STAR Randomness web service
+
+use axum::extract::{Json, State};
+use axum::http::StatusCode;
+use axum::{routing::get, routing::post, Router};
+use base64::prelude::{Engine as _, BASE64_STANDARD as BASE64};
+use serde::{Deserialize, Serialize};
+use tracing::{debug, info};
+
+use ppoprf::ppoprf;
+use std::sync::{Arc, RwLock};
+
+/// Internal state of the OPRF service
+struct OPRFServer {
+    /// oprf implementation
+    server: ppoprf::Server,
+    /// currently-valid randomness epoch
+    epoch: u8,
+}
+
+/// Shareable wrapper around the server state
+type OPRFState = Arc<RwLock<OPRFServer>>;
```

Comment:
should we be using `tokio::sync::RwLock` instead?

Reply:
I don't think so. That's only important if the lock is held across an `await` and I tried to make sure that wasn't happening. The request handlers take a `read()` lock at the start, but don't do anything async in their body. The epoch_update_loop takes a `write()` lock after sleeping, but it should be dropped at the end of the iteration.

Reply:
it seems the lock is held in the async request handlers though? wouldn't that mean that the request handlers will fail to yield to the runtime (and block the whole worker thread instead) if a write lock is held?

Reply:
I think that's right. But the request handlers can't complete while the write lock is held, so blocking is fine. If we yielded instead, maybe we could queue more futures but we wouldn't really be making any progress.

If we're running single-threaded, there can't be any lock contention because only one function can run at a time. It also can't deadlock because the lock guards are dropped before any await. If we're running multi-threaded there can be contention, so individual threads will be blocked up, but we still can't deadlock because each on is separate. Rust actually enforces this: if you try to add an `await` somewhere the lock guard is in scope the compiler will report an `error: future cannot be sent between threads safely`.

I'm trusting here that the scheduler will arrange things to the write lock makes progress under contention. The docs are [completely unhelpful](https://doc.rust-lang.org/std/sync/struct.RwLock.html) here, saying it's up to the operating system, but it doesn't seem like a `RwLock` is useful at all if that's not the case?

I think the deadlock example in the docs might be relying on the shadowed read guard staying alive. We don't to that, since ours are dropped before another is requested, whether through linear or async code.

Reply:
It looks like deadlock could be a concern, if there was an opportunity for `randomness` or `info` to yield (i.e. via an await). If one of the request handlers acquires a read lock and yields to the runtime by awaiting, and another request handler gets executed on the same thread, then it would attempt to acquire a read lock while a write lock is potentially waiting to be acquired, right? Currently, there are no awaits in either handlers so this does not apply right now, but could occur in the future with code changes.

> Rust actually enforces this: if you try to add an await somewhere the lock guard is in scope the compiler will report an error: future cannot be sent between threads safely.

It seems this wouldn't be the case with `tokio::sync::RwLock` since it's guards implement `Send`, right? Are we limiting ourselves by not using the tokio lock? If we used it, then we could await anywhere.

Reply:
On the issue of whether we can have writer starvation, I'm [informed](https://hachyderm.io/@Mara/109984712062723989) that [since 1.62](https://github.com/rust-lang/rust/pull/95801/) `std::sync::RwLock` [prioritizes writers](https://github.com/rust-lang/rust/blob/1.62.1/library/std/src/sys/unix/locks/futex_rwlock.rs#L256-L257) on Linux (and BSD). So it works as expected and is fine for our uses here.